File: vlc-android/src/org/videolan/vlc/android/VlcCrashHandler.java
package org.videolan.vlc.android;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.io.Writer;
import java.lang.Thread.UncaughtExceptionHandler;

import android.text.format.DateFormat;
import android.util.Log;

public class VlcCrashHandler implements UncaughtExceptionHandler {

    private static final String TAG = "VLC/VlcCrashHandler";

    private UncaughtExceptionHandler defaultUEH;

    public VlcCrashHandler() {
        this.defaultUEH = Thread.getDefaultUncaughtExceptionHandler();
    }

    @Override
    public void uncaughtException(Thread thread, Throwable ex) {

        final Writer result = new StringWriter();
        final PrintWriter printWriter = new PrintWriter(result);
        ex.printStackTrace(printWriter);
        String stacktrace = result.toString();
        printWriter.close();

        Log.e(TAG, stacktrace);
        writeLog(stacktrace, "//sdcard//vlc_crash");
        writeLogcat("//sdcard//vlc_logcat");

        defaultUEH.uncaughtException(thread, ex);
    }

    private void writeLog(String log, String name) {
        CharSequence timestamp = DateFormat.format("yyyyMMdd_kkmmss", System.currentTimeMillis());
        String filename = name + "_" + timestamp + ".log";

        try {
            FileOutputStream stream = new FileOutputStream(filename);
            OutputStreamWriter output = new OutputStreamWriter(stream);
            BufferedWriter bw = new BufferedWriter(output);

            bw.write(log);
            bw.newLine();

            bw.close();
            output.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void writeLogcat(String name) {
        CharSequence timestamp = DateFormat.format("yyyyMMdd_kkmmss", System.currentTimeMillis());
        String filename = name + "_" + timestamp + ".log";
        String[] args = { "logcat", "-v", "time", "-d" };

        try {
            Process process = Runtime.getRuntime().exec(args);
            InputStreamReader input = new InputStreamReader(
                    process.getInputStream());
            OutputStreamWriter output = new OutputStreamWriter(
                    new FileOutputStream(filename));
            BufferedReader br = new BufferedReader(input);
            BufferedWriter bw = new BufferedWriter(output);
            String line;

            while ((line = br.readLine()) != null) {
                bw.write(line);
                bw.newLine();
            }

            bw.close();
            output.close();
            br.close();
            input.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: application/resources/src/main/java/org/videolan/resources/util/VLCCrashHandler.kt
/*****************************************************************************
 * VlcCrashHandler.java
 *
 * Copyright © 2012 VLC authors and VideoLAN
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 */

package org.videolan.resources.util

import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Environment
import android.text.format.DateFormat
import android.util.Log
import org.videolan.resources.AppContextProvider
import org.videolan.tools.CloseableUtils
import org.videolan.tools.Logcat
import java.io.*
import java.lang.Thread.UncaughtExceptionHandler

private const val TAG = "VLC/VlcCrashHandler"

class VLCCrashHandler : UncaughtExceptionHandler {

    private val defaultUEH: UncaughtExceptionHandler? = Thread.getDefaultUncaughtExceptionHandler()

    override fun uncaughtException(thread: Thread, ex: Throwable) {
        defaultUEH?.uncaughtException(thread, saveLog(ex))
    }

    companion object {
        /**
         * Saves a [Throwable] stack trace in a crash file after having some useful info appended.
         * Also save a logcat file besides it
         * /!\ Will trigger the crash reporter in beta!
         * It's called by the [UncaughtExceptionHandler] but is useful in case we want to workaround a crash
         * but still get the trace in the crash reporter
         *
         * @param ex: the [Throwable] to log
         * @return the [Throwable] with versions appended
         */
        fun saveLog(ex: Throwable, watermark:String = ""):Throwable {
            val result = StringWriter()
            val printWriter = PrintWriter(result)

            // Inject some info about android version and the device, since google can't provide them in the developer console
            val trace = ex.stackTrace
            val trace2 = arrayOfNulls<StackTraceElement>(trace.size + if (watermark.isNotEmpty()) 4  else 3)
            System.arraycopy(trace, 0, trace2, 0, trace.size)
            trace2[trace.size + 0] = StackTraceElement("Android", "MODEL", android.os.Build.MODEL, -1)
            trace2[trace.size + 1] = StackTraceElement("Android", "VERSION", android.os.Build.VERSION.RELEASE, -1)
            trace2[trace.size + 2] = StackTraceElement("Android", "FINGERPRINT", android.os.Build.FINGERPRINT, -1)
           if (watermark.isNotEmpty()) trace2[trace.size + 3] = StackTraceElement("VLC", "Watermark", watermark, -1)
            ex.stackTrace = trace2

            ex.printStackTrace(printWriter)
            val stacktrace = result.toString()
            printWriter.close()
            Log.e(TAG, stacktrace)

            // Save the log on SD card if available
            if (Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED) {
                writeLog(stacktrace, AppContextProvider.appContext.getExternalFilesDir(null)!!.absolutePath + "/vlc_crash")
                writeLogcat(AppContextProvider.appContext.getExternalFilesDir(null)!!.absolutePath + "/vlc_logcat")
            }
            return ex
        }

        /**
         * Writes a log in a file
         *
         * @param log: the log string to write
         * @param name: the file name to write into
         */
        private fun writeLog(log: String, name: String) {
            val timestamp = DateFormat.format("yyyyMMdd_kkmmss", System.currentTimeMillis())
            val filename = name + "_" + timestamp + ".log"

            val stream: FileOutputStream
            try {
                stream = FileOutputStream(filename)
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
                return
            }

            val output = OutputStreamWriter(stream)
            val bw = BufferedWriter(output)

            val version = try {
                val pInfo: PackageInfo = AppContextProvider.appContext.packageManager.getPackageInfoCompat(AppContextProvider.appContext.packageName, 0)
                pInfo.versionName
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
            }

            try {
                bw.write("App version: $version\r\n")
                bw.write(log)
                bw.newLine()
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                CloseableUtils.close(bw)
                CloseableUtils.close(output)
            }
        }

        /**
         * Write the current log in a file
         *
         * @param name: the file name to use to save the logcat
         */
        private fun writeLogcat(name: String) {
            val timestamp = DateFormat.format("yyyyMMdd_kkmmss", System.currentTimeMillis())
            val filename = name + "_" + timestamp + ".log"
            try {
                Logcat.writeLogcat(filename)
            } catch (e: IOException) {
                Log.e(TAG, "Cannot write logcat to disk")
            }

        }
    }
}
